tests/MongoDB.Analyzer.Tests/Infrastructure/TestCasesRunner.cs

```diff
@@ -58,35 +61,51 @@ public static async Task<DiagnosticTestCaseResult> RunTestCase(DiagnosticTestCas
 
     private static async Task<IDictionary<string, DiagnosticTestCaseResult>> ExecuteTestCases(TestsBundleKey testsBundleKey)
     {
+        var isPoco2JsonTest = testsBundleKey.TestFileName.Contains(s_poco2JsonTestPath);
+
         var diagnostics = await DiagnosticsAnalyzer.Analyze(
             testsBundleKey.TestFileName,
             testsBundleKey.DriverVersion,
             testsBundleKey.LinqVersion);
 
         var diagnosticsAndMethodNodes = diagnostics
-            .Where(d => DiagnosticRulesConstants.AllRules.Contains(d.Descriptor.Id))
-            .Select(d =>
-                (Diagnostic: d,
-                 MethodNode: d.Location.SourceTree.GetRoot()
-                 .FindNode(d.Location.SourceSpan)
-                 .Ancestors()
-                 .OfType<MethodDeclarationSyntax>()
-                 .FirstOrDefault()))
-            .Where(d => d.MethodNode != null)
-            .ToArray();
+                                            .Where(d => DiagnosticRulesConstants.AllRules.Contains(d.Descriptor.Id))
+                                            .Select(d =>
+                                                (Diagnostic: d,
+                                                 MethodNode: FindMethodNode(d, isPoco2JsonTest)))
+                                            .Where(d => d.MethodNode != null)
+                                            .ToArray();
 
         var result = diagnosticsAndMethodNodes
-            .GroupBy(pair => pair.MethodNode.Identifier.Text)
-            .ToDictionary(
-                group => group.Key,
-                group => new DiagnosticTestCaseResult(
-                    group.Key,
-                    GetMethodLocation(group.FirstOrDefault().MethodNode),
-                    group.Select(pair => pair.Diagnostic).ToArray()));
+                    .GroupBy(pair => pair.MethodNode.Identifier.Text)
+                    .ToDictionary(
+                        group => group.Key,
+                        group => new DiagnosticTestCaseResult(
+                            group.Key,
+                            GetMethodLocation(group.FirstOrDefault().MethodNode),
+                            group.Select(pair => pair.Diagnostic).ToArray()));
 
         int GetMethodLocation(MethodDeclarationSyntax node) =>
             node.DescendantNodes().OfType<BlockSyntax>().First().GetLocation().GetLineSpan().StartLinePosition.Line;
 
         return result;
     }
+
+    private static MethodDeclarationSyntax FindMethodNode(Diagnostic diagnostic, bool isPoco2JsonTest)
+    {
+        if (isPoco2JsonTest)
+        {
+            var syntaxRoot = diagnostic.Location.SourceTree.GetRoot();
+            var classNode = syntaxRoot.FindNode(diagnostic.Location.SourceSpan) as ClassDeclarationSyntax;
+
+            var methodNode = syntaxRoot.DescendantNodes().OfType<MethodDeclarationSyntax>().FirstOrDefault(method => method.Identifier.ValueText == classNode.Identifier.ValueText);
```

Comment:
Try to optimize the style of this method, for example seems that GetRoot and .FindNode(diagnostic.Location.SourceSpan) are invoked in both branches.

Reply:
Done